.github/workflows/all.yml

```diff
@@ -0,0 +1,319 @@
+    runs-on: ${{ matrix.platform }}
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v2
+      - name: Install Go
+        uses: actions/setup-go@v2
+        with:
+          go-version: ${{ matrix.go-version }}
+      - name: Run tests
+        run: |
+          set -x
+          go version
+          export GOMAXPROCS=2
+          args=("-race")
+          # Run with less concurrency on Windows to minimize flakiness.
+          if [[ "${{ matrix.platform }}" == windows* ]]; then
+            unset args[0]
+            export GOMAXPROCS=1
+          fi
+          go test "${args[@]}" -timeout 800s ./...
+
+  test-coverage:
+    strategy:
+      matrix:
+        go-version: [1.15.x]
+        platform: [ubuntu-latest, windows-latest]
+    runs-on: ${{ matrix.platform }}
+    env:
+      CODECOV_TOKEN: ${{ secrets.CODECOV_TOKEN }}
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v2
+      - name: Install Go
+        uses: actions/setup-go@v2
+        with:
+          go-version: ${{ matrix.go-version }}
+      - name: Run tests with code coverage
+        run: |
+          go version
+          export GOMAXPROCS=2
+          args=("-race")
+          # Run with less concurrency on Windows to minimize flakiness.
+          if [[ "${{ matrix.platform }}" == windows* ]]; then
+            unset args[0]
+            export GOMAXPROCS=1
+          fi
+          echo "mode: set" > coverage.txt
+          for pkg in $(go list ./... | grep -v vendor); do
+              list=$(go list -test -f  '{{ join .Deps  "\n"}}' $pkg | grep github.com/loadimpact/k6 | grep -v vendor || true)
+              if [ -n "$list" ]; then
+                  list=$(echo "$list" | cut -f1 -d ' ' | sort -u | paste -sd, -)
+              fi
+
+              go test "${args[@]}" -timeout 800s --coverpkg="$list" -coverprofile=$(echo $pkg | tr / -).coverage $pkg
+          done
+          grep -h -v "^mode:" *.coverage >> coverage.txt
+          rm -f *.coverage
+      - name: Upload coverage to Codecov
+        run: |
+          platform="${{ matrix.platform }}"
+          bash <(curl --fail -s https://codecov.io/bash) -F "${platform%%-*}"
+      - name: Generate coverage HTML report
+        run: go tool cover -html=coverage.txt -o coverage.html
+      - name: Upload coverage report
+        uses: actions/upload-artifact@v2
+        with:
+          name: test-coverage-report-${{ matrix.platform }}
+          path: coverage.html
+
+  configure:
+    runs-on: ubuntu-latest
+    if: startsWith(github.ref, 'refs/tags/v')
+    outputs:
+        version: ${{ steps.get_version.outputs.version }}
+    steps:
+      - name: Get version
+        id: get_version
+        run: |
+          VERSION="${GITHUB_REF##*/}"
+          echo "VERSION=${VERSION}"
+          echo "::set-output name=version::${VERSION}"
+
+  build:
+    runs-on: ubuntu-latest
+    needs: [deps, lint, test]
+    if: startsWith(github.ref, 'refs/tags/v')
```

Comment:
I think this is sub-optimal - the binaries and packages should be built on every k6 version, but they should just be published on `v*` tags.

Reply:
Hhmm sure, I'll change it.

Reply:
So I like the idea of building binaries for all PRs, as that would make testing them much easier.

But there's currently a couple of minor problems with this:
- The `$VERSION` for these builds is the short Git SHA, which [fails the semantic version check for the RPM package](https://github.com/imiric/k6/runs/1225740444). I can change this to be similar to the `git describe` output, but RPM versions are also [incompatible with semver](https://github.com/semver/semver/issues/145), so we'd have to use tildes instead of dashes as suggested there, e.g. `<latest Git tag>~<commit SHA>`.
- The Windows .msi and Chocolatey packages are built in `publish-windows`, which should probably be done in a separate `build` job. But even then I know that the `.msi` package also has issues with semver, and we'd need a different workaround. You probably didn't mean to imply building Windows packages as well on every PR and on `master`, and since we're building the binary anyway, they're probably not needed.

So maybe to simplify this, what do you think about removing all packaging from `build-release.sh` (essentially reverting most of the changes I've done here...), and only building binaries and uploading them as .tar.gz and .zip on each PR/`master`? Leaving packaging exclusively for when building a `v*` tag, and probably keep using the same `gen-packages.sh` script as we do now.

Reply:
:disappointed: Let's leave this for now then... In the interests of the perfect not being an enemy of the good, we can handle it as a separate PR after this is merged. I'd be much easier to iterate the CI once we no longer have CircleCI and AppVeyor anyway.

Reply:
It wouldn't be much work to revert some of these changes, upload binaries only on all commits, and keep using `gen-packages.sh` for `v*` tag builds only, though. But OK, it can wait for later.

Reply:
why do want to *upload* binaries on *all* commits?

Reply:
I'd like to build all the packages, so we know that is not broken when it comes to release.

I can also see _some_ value in saving them and uploading them as artifacts, but that would probably quickly get out of hand and we'd run against some limit if we don't clean them. So, I am not certain if we should do this, probably not. It's not difficult to build k6 yourself from a branch you want to test.

Reply:
@MStoykov Because it would make them available for quick testing, as we've recently had to do with the DNS PRs. If we're building them anyway, I don't see a problem with uploading them as build artifacts, unless GH starts charging us or something.

Reply:
@imiric the storage limits seems super low, so I doubt we'll save all artifacts: https://docs.github.com/en/free-pro-team@latest/github/setting-up-and-managing-billing-and-payments-on-github/about-billing-for-github-actions

Reply:
Ah, yeah, we'd run out of free storage quickly in that case. OK, then let's leave building/publishing for version tags only for now.